### PR TITLE
MWPW-168589: Placeholders in M@S

### PR DESCRIPTION
**Pull request by Axelcureno**
This PR introduces an all-new Placeholder management experience in Merch at Scale Studio that enables CRUD operations using Odin as the data source.
Other improvements included in this PR are:

- code enhancements and cleanup.
- Refactored, more resilient and consistent routing navigation.
- Fixes and expands test coverage, including a new mock service that covers more files.
- Lint fixes.

Resolve  https://jira.corp.adobe.com/browse/MWPW-168589

Test URLs:
- Before: https://main--mas--adobecom.aem.live/studio.html?page=placeholders#path=nala&locale=en_US
- After: https://mwpw-168589--mas--adobecom.aem.live/studio.html?page=placeholders#path=nala&locale=en_US

### CODE REVIEW

File: studio/src/mas-side-nav.js [hunk 1]
```diff
@@ -77,19 +81,17 @@ class MasSideNav extends LitElement {
                     label="Home"
                     value="home"
                     @click="${navigateToPage('welcome')}"
-                    selected
+                    ?selected=${this.currentPage.value === 'welcome'}
```
Comment (yesil) [inline studio/src/mas-side-nav.js:84]: Please use `Store.page`. We discussed pros & cons of using StoreController for accessing data and the outcome was to use `Store.` api directly. 
Thus, we keep also automatic code completion.

File: studio/src/mas-side-nav.js [hunk 2]
```diff
@@ -77,19 +81,17 @@ class MasSideNav extends LitElement {
...
-                </sp-sidenav-item>
-
-                <sp-sidenav-item label="Reporting" value="reporting">
-                    <sp-icon-graph-bar-vertical
-                        slot="icon"
-                    ></sp-icon-graph-bar-vertical>
+                <sp-sidenav-item
+                    label="Placeholders"
+                    value="placeholders"
+                    @click="${navigateToPage('placeholders')}"
```
Comment (yesil) [inline studio/src/mas-side-nav.js:91]: time to make these values (placeholders, content, welcome) a constant?

File: studio/src/store.js [hunk 3]
```diff
@@ -97,105 +109,94 @@ export function navigateToPage(value) {
...
+            updates[key] = value;
+            shouldUpdateStore = true;
+        } else if (defaultValue !== undefined) {
+            const defaultForKey = Array.isArray(defaultValue)
+                ? defaultValue
+                : typeof defaultValue === 'object' && defaultValue !== null
+                  ? defaultValue[key]
+                  : defaultValue;
+
+            if (defaultForKey !== undefined) {
```
Comment (yesil) [inline studio/src/store.js:152]: I also changed this part in my PR #196, I wonder what was the reason you changed this?
in mine I noticed, that `syncFromHash` was causing extra search calls, also breaking the loading indicator.

File: studio/src/store.js [hunk 4]
```diff
@@ -97,105 +109,94 @@ export function navigateToPage(value) {
...
-        }
-        setHashParams(hashParams, normalizedValue);
-        window.location.hash = hashParams.toString();
+    if (shouldUpdateStore) {
+        store.set((prev) => ({ ...prev, ...updates }));
     }
 
-    // Initialize
-    syncFromHash();
+    store.subscribe((value) => {
```
Comment (yesil) [inline studio/src/store.js:163]: ❤️

File: studio/src/store.js [hunk 5]
```diff
@@ -209,11 +210,47 @@ export function unlinkStoreFromHash(store) {
...
-// When the query param is populated or changes, switch to the 'content' page.
 Store.search.subscribe((value, oldValue) => {
     if (
         (!oldValue.query && value.query) ||
         (Boolean(value.query) && value.query !== oldValue.query)
     )
         Store.page.set('content');
 });
+
+document.addEventListener('DOMContentLoaded', () => {
```
Comment (yesil) [inline studio/src/store.js:221]: store.js is growing too fast.
can we move this in a separate file?
name suggestions: router.js, deeplink.js.

File: studio/src/studio.js [hunk 6]
```diff
@@ -52,10 +53,24 @@ class MasStudio extends LitElement {
...
 
     disconnectedCallback() {
         super.disconnectedCallback();
+        window.removeEventListener(
+            'popstate',
+            this.updatePageFromUrl.bind(this),
+        );
+    }
+
+    updatePageFromUrl() {
```
Comment (yesil) [inline studio/src/studio.js:68]: initialization of the store from the URL should be centralized at a single place. Otherwise it will be very hard to troubleshoot.

File: studio/test/aem/aem.test.js [hunk 7]
```diff
@@ -1,8 +1,106 @@
...
+
+        fetchStub = sinon.stub(window, 'fetch');
+    });
+
+    afterEach(() => {
+        sinon.restore();
+    });
+
+    describe('fragments.create', () => {
+        it('should create fragment successfully', async () => {
```
Comment (yesil) [inline studio/test/aem/aem.test.js:47]: I don't think we can test the creation.

File: studio/test/aem/aem.test.js [hunk 8]
```diff
@@ -1,8 +1,106 @@
...
+                await aem.sites.cf.fragments.create({}, '/test/path');
+                expect.fail('Should have thrown an error');
+            } catch (err) {
+                expect(err.message).to.include('Bad Request');
+            }
+        });
+    });
+
+    describe('fragments.publish', () => {
+        it('should publish fragment successfully', async () => {
```
Comment (yesil) [inline studio/test/aem/aem.test.js:79]: to test aem.js, I recommend to test the payload of the fetch requests being send.

File: studio/test/aem/mas-placeholders.test.js [hunk 9]
```diff
@@ -0,0 +1,300 @@
...
+     * Initializes the component with test data
+     */
+    beforeEach(() => {
+        // Create repository mock
+        mockRepository = {
+            searchPlaceholders: sinon.stub().resolves([
+                {
+                    key: 'test-key',
+                    value: 'test-value',
+                    _fragment: {
```
Comment (yesil) [inline studio/test/aem/mas-placeholders.test.js:31]: please do not start variables with `_`.

File: studio/test/aem/mas-placeholders.test.js [hunk 10]
```diff
@@ -0,0 +1,300 @@
...
+                    ],
+                },
+            },
+        ];
+
+        /**
+         * Mock Store object that simulates the reactive data store
+         * Includes placeholders, search, filters and page state
+         */
+        window.Store = {
```
Comment (yesil) [inline studio/test/aem/mas-placeholders.test.js:86]: maybe you can create a store.js in mocks folder?

File: studio/test/aem/mas-placeholders.test.js [hunk 11]
```diff
@@ -0,0 +1,300 @@
...
+                        { name: 'key', values: ['test-key'] },
+                        { name: 'value', values: ['Updated Value'] },
+                    ],
+                });
+
+                // Reset editing state
+                this.editingPlaceholder = null;
+
+                // Emit success toast
+                window.Events.toast.emit({
```
Comment (yesil) [inline studio/test/aem/mas-placeholders.test.js:271]: the events such as `placeholderAdded` do not seem to be leveraged here.

File: studio/test/mocks/aem-service-mock.js [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+ */
+export class AEMServiceMock {
+  constructor(config = {}) {
+    this.cfFragmentsUrl = config.baseUrl || '/mock/content/dam/fragments';
+    this.headers = { Authorization: 'Bearer mock-token' };
+    
+    // Define fragment models
+    this._fragmentModels = new Map();
+    this._fragmentModels.set('test-model', {
+      id: 'test-model',
```
Comment (yesil) [inline studio/test/mocks/aem-service-mock.js:12]: please do not start variables with underscore.

File: studio/test/mocks/aem-service-mock.js [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+      method.allArgs = [];
+      method.nextResponse = null;
+      method.shouldReject = false;
+      method.rejectReason = null;
+      return method;
+    };
+    
+    return method;
+  }
+}
```
Comment (yesil) [inline studio/test/mocks/aem-service-mock.js:138]: EOF

File: studio/test/mocks/aem-service-mock.js
Comment (yesil) [inline studio/test/mocks/aem-service-mock.js:1]: mocks should have the same name as the module they are mocking for ease of discoverability.

File: studio/test/mocks/aem-service-mock.js [hunk 14]
```diff
@@ -0,0 +1,138 @@
+/**
+ * Complete AEM service mock for testing
+ */
+export class AEMServiceMock {
```
Comment (yesil) [inline studio/test/mocks/aem-service-mock.js:4]: where is this used? I would like to avoid reviewing AI generated code that is not used.

File: studio/src/mas-placeholders.js [hunk 15]
```diff
@@ -0,0 +1,1479 @@
...
+        return null;
+    }
+
+    const surface = folderPath;
+    const dictionaryPath = `${ROOT_PATH}/${surface}/${locale}/dictionary`;
+    return dictionaryPath;
+}
+
+class MasPlaceholders extends LitElement {
+    static styles = css`
```
Comment (yesil) [inline studio/src/mas-placeholders.js:26]: can you move styles to a css.js file?  when there are too many styles, it is hard to edit the file.

File: studio/src/mas-placeholders.js [hunk 16]
```diff
@@ -0,0 +1,1479 @@
...
+            value: '35%',
+            locale: '10%',
+            status: '10%',
+            updatedBy: '10%',
+            updatedAt: '10%',
+            action: '5%',
+        };
+
+        this.selectedFolder = new StoreController(
+            this,
```
Comment (yesil) [inline studio/src/mas-placeholders.js:397]: can you use a single `ReactiveController` instead?

reactiveController = new ReactiveController(this, [...stores])

File: studio/src/mas-placeholders.js [hunk 17]
```diff
@@ -0,0 +1,1479 @@
...
+        const currentPage = Store.page.get();
+        if (currentPage !== 'placeholders') {
+            Store.page.set('placeholders');
+        }
+
+        const masRepository = this.repository;
+        if (masRepository) {
+            if (this.selectedFolder.value?.path) {
+                Store.placeholders.list.loading.set(true);
+                setTimeout(() => {
```
Comment (yesil) [inline studio/src/mas-placeholders.js:459]: why setTimeout is needed?

File: studio/src/mas-placeholders.js [hunk 18]
```diff
@@ -0,0 +1,1479 @@
...
+
+            const savedFragment =
+                await this.repository.saveDictionaryFragment(updatedFragment);
+
+            const updatedPlaceholder = {
+                ...placeholder,
+                key: this.editedKey,
+                value: this.editedValue,
+                status: savedFragment.status || 'Draft',
+                _fragment: savedFragment,
```
Comment (yesil) [inline studio/src/mas-placeholders.js:696]: please respect coding convention, we do not start variables with underscore.

File: studio/src/mas-placeholders.js [hunk 19]
```diff
@@ -0,0 +1,1479 @@
...
+            this.activeDropdown = null;
+
+            await this.repository.deleteDictionaryFragment(fragmentData);
+
+            const updatedPlaceholders = this.placeholdersData.value.filter(
+                (p) => p.key !== key,
+            );
+            Store.placeholders.list.data.set(updatedPlaceholders);
+
+            setTimeout(() => {
```
Comment (yesil) [inline studio/src/mas-placeholders.js:866]: why `setTimeout` is needed?

File: studio/src/mas-placeholders.js [hunk 20]
```diff
@@ -0,0 +1,1479 @@
...
+        setTimeout(() => {
+            if (!this.showCreateModal) {
+                this.showCreateModal = true;
+                this.requestUpdate();
+            }
+        }, 100);
+    }
+
+    handleExport() {
+        // ToDo: Implement export functionality
```
Comment (yesil) [inline studio/src/mas-placeholders.js:948]: is this needed?

File: studio/src/mas-placeholders.js [hunk 21]
```diff
@@ -0,0 +1,1479 @@
...
+        const newLocale = event.detail.locale;
+        Store.filters.value.locale = newLocale;
+        this.selectedLocale.value = newLocale;
+        Store.placeholders.list.loading.set(true);
+
+        if (this.repository) {
+            this.repository.searchPlaceholders();
+        }
+
+        this.requestUpdate();
```
Comment (yesil) [inline studio/src/mas-placeholders.js:1278]: nit: this should not be needed, as it will refresh when one of the observed stores is updated.

**Review by yesil — CHANGES_REQUESTED** [review]

File: studio/src/store.js [hunk 22]
```diff
@@ -8,7 +8,11 @@ import {
     setHashParams,
```
Comment (st-angelo-adobe) [inline studio/src/store.js:8]: There are some unused methods being imported

File: studio/src/mas-placeholders.js [hunk 23]
```diff
@@ -0,0 +1,1479 @@
...
+            action: '5%',
+        };
+
+        this.selectedFolder = new StoreController(
+            this,
+            Store.search || new ReactiveStore({}),
+        );
+        this.selectedLocale = new StoreController(
+            this,
+            Store.filters?.locale || new ReactiveStore('en_US'),
```
Comment (st-angelo-adobe) [inline studio/src/mas-placeholders.js:402]: We shouldn't be creating new reactive stores anywhere outside the `store.js` file. Instead we should use the existing one, so we simply subscribe to the `filters` store, and to the `folders.data` store.

File: studio/src/mas-placeholders.js [hunk 24]
```diff
@@ -0,0 +1,1479 @@
...
+                        </div>
+                    </div>
+                </div>
+            </div>
+        `;
+    }
+
+    handleLocaleChange(event) {
+        const newLocale = event.detail.locale;
+        Store.filters.value.locale = newLocale;
```
Comment (st-angelo-adobe) [inline studio/src/mas-placeholders.js:1270]: We should NEVER manually set a store's value like this, because then the subscribers don't get notified.
We should only update the value of the store using `set`.

File: studio/src/store.js (hunk 3 above)
Comment (Axelcureno) [reply studio/src/store.js:152]: In general I made the handling of search params more robust by implementing standardized URL parsing, supports complex data structures via JSON serialization and enhances maintainability with cleaner, more efficient code organization. This particular section improves performance by avoiding multiple reactive updates when setting several default values at once (like the issue you described).

File: studio/src/store.js [hunk 25]
```diff
@@ -97,105 +109,94 @@ export function navigateToPage(value) {
...
+            if (currentParams.get(key) !== stringValue) {
+                currentParams.set(key, stringValue);
+                hasChanges = true;
+            }
+        }
 
-    // Subscribe
-    window.addEventListener('hashchange', syncFromHash);
-    store.subscribe(syncToHash);
-    store.setMeta('hashLink', { from: syncFromHash, to: syncToHash });
```
[the remark this answers is not in the document]
Comment (Axelcureno) [reply studio/src/store.js:198]: I revamped the URL parameter handling mechanism.
The new implementation delivers significant advantages through five key improvements:
1. **Better Data Structure Support**: Properly handles complex objects and arrays through JSON serialization, ensuring accurate preservation of user settings in URLs.
2. **Performance Optimization**: Collects all parameter changes and applies them in a single update, eliminating unnecessary re-renders and improving page load speed.
3. **Smarter History Management**: Uses the History API to update the current browser history entry rather than creating multiple entries, providing a more intuitive navigation experience.
4. **Consistent Default Values**: Implements a clear, predictable system for applying default values when parameters are missing, ensuring a reliable user experience.
5. **Enhanced Shareability**: Makes URLs perfectly capture application state, allowing users to bookmark specific views or share exact configurations with colleagues.

File: studio/test/mocks/aem-service-mock.js (hunk 14 above)
Comment (Axelcureno) [reply studio/test/mocks/aem-service-mock.js:4]: Sorry, I forgot to remove this.

File: studio/src/mas-side-nav.js (hunk 1 above)
Comment (Axelcureno) [reply studio/src/mas-side-nav.js]: Addressed

File: studio/src/mas-side-nav.js (hunk 2 above)
Comment (Axelcureno) [reply studio/src/mas-side-nav.js]: done

File: studio/src/store.js (hunk 5 above)
Comment (Axelcureno) [reply studio/src/store.js]: I used your suggestion `router.js` and moved navigation-related stuff there.

File: studio/src/mas-placeholders.js (hunk 17 above)
Comment (Axelcureno) [reply studio/src/mas-placeholders.js]: removed

File: studio/src/mas-placeholders.js (hunk 18 above)
Comment (Axelcureno) [reply studio/src/mas-placeholders.js]: fixed

File: studio/src/mas-placeholders.js (hunk 19 above)
Comment (Axelcureno) [reply studio/src/mas-placeholders.js]: removed

File: studio/src/mas-placeholders.js (hunk 20 above)
Comment (Axelcureno) [reply studio/src/mas-placeholders.js]: It will in the future but I have removed it in the meantime. You may noticed I also removed other stuff like this that is not used.

File: studio/src/mas-placeholders.js (hunk 21 above)
Comment (Axelcureno) [reply studio/src/mas-placeholders.js]: Removed

File: studio/src/mas-placeholders.js (hunk 23 above)
Comment (Axelcureno) [reply studio/src/mas-placeholders.js]: 🫠 removed

**Review by yesil — COMMENTED** [review]
reset filters button is broken.

image

File: studio/src/events.js [hunk 26]
```diff
@@ -3,6 +3,10 @@ import MasEvent from './reactivity/mas-event.js';
 const Events = {
     toast: new MasEvent(),
     fragmentAdded: new MasEvent(),
+    placeholderAdded: new MasEvent(),
```
[the remark this answers is not in the document]
Comment (Axelcureno) [reply studio/src/events.js]: removed

**Comment by Axelcureno** [comment]
> reset filters button is broken.
> 
> image

@yesil good callout, thanks. I have fixed it by doing the following:
- Made sure both data stores (search and filter) are reset when clicking "reset filters" (previously only one was being updated).
- Added code to explicitly clear filter parameters from the URL (this was causing the issue).
- Improved how the app detects filter changes by adding special handling for empty tag arrays and **subscribing to filter changes directly**.

**Review by yesil — APPROVED** [review]

**Review by 3ch023 — CHANGES_REQUESTED** [review]
Hi @Axelcureno , was checking it today and it looks great!
one small issue i found is that 'index' placeholder containing all other placeholders is not updated when i add a new placeholder.
@npeltier can help with details on how its supposed to work, could you sync with him pls and address?
this is my Milo test page with M@S card with 2 placeholders:
https://stage--milo--adobecom.aem.page/drafts/mariia/document5

studio: https://mas.adobe.com/studio.html#path=sandbox&query=a61c723c-3c07-43d5-a082-97609a4da4e7&page=content
you can see {{cai-default}} works, but my new {{mariia-placeholder}} not yet.
i assume because https://odin.adobe.com/adobe/sites/fragments?path=/content/dam/mas/sandbox/en_US/dictionary/index doesn't have a reference to it.

+ i wouldn't check Nala till Ilyases PR on collection goes in. once its merged you can pull latest & recheck

**Review by npeltier — CHANGES_REQUESTED** [review]
Hi @Axelcureno , as discussed offline, when a new placeholder is created for a surface / locale, and saved (before it's even published) you should  
1. check for existence of an index fragment for that surface / locale if, not existing create it,
2. add that new fragment as reference to index
3. publish index (that should publish new fragment too)

i discussed the drawback of having a draft published, but at that stage, the PH is not used so it's good.

File: studio/src/mas-side-nav.js [hunk 27]
```diff
@@ -76,20 +81,18 @@ class MasSideNav extends LitElement {
...
-
-                <sp-sidenav-item label="Promotions" value="promotions">
-                    <sp-icon-promote slot="icon"></sp-icon-promote>
-                </sp-sidenav-item>
-
-                <sp-sidenav-item label="Reporting" value="reporting">
-                    <sp-icon-graph-bar-vertical
-                        slot="icon"
-                    ></sp-icon-graph-bar-vertical>
+                <sp-sidenav-item
```
[the remark this answers is not in the document]
Comment (Axelcureno) [reply studio/src/mas-side-nav.js:89]: added

File: studio/src/mas-top-nav.js [hunk 28]
```diff
@@ -3,9 +3,16 @@ import { LitElement, html, css, until } from 'lit';
 import Store from './store.js';
```
[the remark this answers is not in the document]
Comment (Axelcureno) [reply studio/src/mas-top-nav.js:3]: Which changes?

**Comment by Axelcureno** [comment]
Addressed all your comments @npeltier @3ch023 :)

**Comment by npeltier** [comment]
Hey @Axelcureno current published index still only have 4 entries, see: 
https://odin.adobe.com/adobe/sites/fragments/4218bdb3-8d25-4706-974d-21b1f0f13aa9

**Comment by npeltier** [comment]
i tried to delete / create / publish and it's still the same

**Comment by afmicka** [comment]
@Axelcureno Nala is not able to authenticate because something has changed in the studio url.
As far as i can see you are adding and placing `page=welcome` before the hash (`#`). Also moving `page=content` before it.
On your branch it is `?page=content#path=acom`, and before it was `#page=content&path=acom`. Is that intentional? 
Is yes, some of the test specs and authentication need to be updated to match it and be able to execute the tests.

**Comment by npeltier** [comment]
also @Axelcureno FYI just tested creation of placeholder in an empty dictionary folder use case that didn't work. Not sure we make this part of your PR to be fixed (i think i saw create if not exist) or make it part of initial content script creation (we _will_ need one as placeholder needs permissions setup you won't do in your code). But either way change is requested here too:
- fix the create if not exist,
- remove that code
cc @3ch023 @markovicpredrag1209 @yesil

File: studio/src/mas-repository.js [hunk 29]
```diff
@@ -314,36 +339,57 @@ export class MasRepository extends LitElement {
...
+
     async createFragment(fragmentData) {
         try {
-            Events.toast.emit({
-                variant: 'info',
-                content: 'Creating fragment...',
-            });
+            const isPlaceholder = fragmentData.data && (fragmentData.data.key !== undefined || 
+                                                       fragmentData.parentPath?.includes('/dictionary/'));
+            this.showToast(isPlaceholder ? 'Creating placeholder...' : 'Creating fragment...');
```
Comment (npeltier) [inline studio/src/mas-repository.js:364]: i think you can just leave it as create Fragment, that's fine

File: studio/src/mas-repository.js [hunk 30]
```diff
@@ -355,35 +401,89 @@ export class MasRepository extends LitElement {
...
+    async saveFragment(fragment, options = {}) {
+        const { isInEditStore = true } = options;
+        const fragmentToSave = isInEditStore ? this.fragmentInEdit : fragment;
+        
+        if (!fragmentToSave) {
+            throw new Error('No fragment provided for saving');
+        }
+        
+        const isDictionaryFragment = fragmentToSave.path?.includes('/dictionary/');
+        this.showToast(isDictionaryFragment ? 'Saving placeholder...' : 'Saving fragment...');
```
Comment (npeltier) [inline studio/src/mas-repository.js:420]: we shouldn't do that either, just saving frgment is fine

File: studio/src/mas-repository.js [hunk 31]
```diff
@@ -355,35 +401,89 @@ export class MasRepository extends LitElement {
...
+            }
+            
+            if (isInEditStore) {
+                this.fragmentStoreInEdit.refreshFrom(savedFragment);
+            }
+            
+            this.showToast(
+                isDictionaryFragment ? 'Placeholder successfully saved.' : 'Fragment successfully saved.',
+                'positive'
             );
```
Comment (npeltier) [inline studio/src/mas-repository.js:473]: please move this all to a different "updatePlaceholderIndex" method you call from placeholder js together within createion code, and fix the payload, right now it's set to fragment id, hence the publish is broken atm

**Review by npeltier — CHANGES_REQUESTED** [review]

File: studio/src/mas-repository.js [hunk 32]
```diff
@@ -492,6 +500,791 @@ export class MasRepository extends LitElement {
...
+            throw error;
+        }
+    }
+
+    /**
+     * Create a new dictionary/placeholder fragment
+     * @param {Object} fragmentData - The fragment data to create
+     * @returns {Promise<Object>} - The created fragment
+     */
+    async createDictionaryFragment(fragmentData) {
```
[the remark this answers is not in the document]
Comment (Axelcureno) [reply studio/src/mas-repository.js]: done

File: studio/src/mas-repository.js [hunk 33]
```diff
@@ -492,6 +500,791 @@ export class MasRepository extends LitElement {
...
+            }
+
+            const fragmentObjectWithStatus = {
+                ...fragmentData,
+                status: 'PUBLISHED'
+            };
+            
+            const newFragment = await this.createDictionaryFragment(fragmentObjectWithStatus);
+            
+            const publishedFragment = await this.publishDictionaryFragment(newFragment);
```
[the remark this answers is not in the document]
Comment (Axelcureno) [reply studio/src/mas-repository.js]: done

File: studio/src/constants.js [hunk 34]
```diff
@@ -110,12 +111,22 @@ export const LOCALE_DEFAULT = 'en_US';
...
+};
+
 export const TAG_STATUS_PUBLISHED = 'mas:status/published';
 export const TAG_STATUS_PUBLISHED_PATH =
     '/content/cq:tags/mas/status/published';
 export const TAG_STATUS_DRAFT = 'mas:status/draft';
 export const TAG_STATUS_DRAFT_PATH = '/content/cq:tags/mas/status/draft';
 
+export const ROOT_PATH = '/content/dam/mas';
+export const DICTIONARY_MODEL_ID = 'L2NvbmYvbWFzL3NldHRpbmdzL2RhbS9jZm0vbW9kZWxzL2RpY3Rpb25uYXJ5';
```
[the remark this answers is not in the document]
Comment (Axelcureno) [reply studio/src/constants.js:127]: it is being used.

File: studio/src/mas-repository.js [hunk 35]
```diff
@@ -314,36 +339,57 @@ export class MasRepository extends LitElement {
...
+            
+            if (!isPlaceholder) {
+                this.showToast('Fragment successfully created.', 'positive');
+            }
+            
             this.operation.set();
             return new FragmentStore(fragment);
         } catch (error) {
-            this.processError(error, 'Failed to create fragment.');
+            this.processError(error, isPlaceholder ? 'Failed to create placeholder.' : 'Failed to create fragment.');
```
[the remark this answers is not in the document]
Comment (Axelcureno) [reply studio/src/mas-repository.js]: done

File: studio/src/mas-repository.js (hunk 29 above)
Comment (Axelcureno) [reply studio/src/mas-repository.js]: done

File: studio/src/mas-repository.js (hunk 30 above)
Comment (Axelcureno) [reply studio/src/mas-repository.js]: done

File: studio/src/mas-repository.js (hunk 31 above)
Comment (Axelcureno) [reply studio/src/mas-repository.js]: done

**Comment by Axelcureno** [comment]
All remaining comments addressed @npeltier

File: nala/studio/studio.spec.js [hunk 36]
```diff
@@ -34,13 +34,14 @@ export default {
             data: {
                 cardid: '0bf35134-e5e4-4664-88d9-4b78203bf625',
             },
-            browserParams: '#path=nala&page=content',
+            browserParams: '?locale=en_US#path=nala&page=content',
```
Comment (afmicka) [inline nala/studio/studio.spec.js:37]: also here. It is expected that the locale automatically gets recognized if user specifies only path or page

File: nala/studio/studio.test.js [hunk 37]
```diff
@@ -127,7 +127,7 @@ test.describe('M@S Studio feature test suite', () => {
         page,
         baseURL,
     }) => {
-        const testPage = `${baseURL}${features[4].path}${miloLibs}`;
+        const testPage = `${baseURL}${features[4].path}${miloLibs}${features[4].browserParams}`;
```
Comment (afmicka) [inline nala/studio/studio.test.js:130]: i believe adding locale param here is also not needed. It you have issues with the tests without it, lets connect and discuss their behaviour.

**Review by npeltier — APPROVED** [review]
ok, works for me :-) this is great, thanks!

i'm still not a big fan of if tests within generic methods, but i guess this can be changed later

**Review by 3ch023 — APPROVED** [review]

File: studio/src/mas-placeholders.js [hunk 38]
```diff
@@ -0,0 +1,1796 @@
...
+                throw new Error('Fragment data is missing or invalid');
+            }
+
+            Store.placeholders.list.loading.set(true);
+
+            const repository = this.repository;
+            const fragmentId = fragmentData.id;
+            const fragmentEndpoint = `${repository.aem.cfFragmentsUrl}/${fragmentId}`;
+
+            const getResponse = await fetch(fragmentEndpoint, {
```
[the remark this answers is not in the document]
Comment (Axelcureno) [reply studio/src/mas-placeholders.js:335]: It was initially in mas-repository but @npeltier asked me to move things here 🥲

**Comment by Axelcureno** [comment]
@Roycethan my notes regarding testing of placeholders.
1. I have removed offer selector tool from placeholder creation, that is not in scope for now.
2. I was able to modify links on an existing placeholder (added/removed link, changed style) with no issues, please try again and if the issue persists let me know what placeholder you are trying with.
3. Please test only in locales that already have placeholders for now. This is happening because not all locales have the folder created in Odin. @npeltier said this will be handled separately in another ticket, a script to create the folders in Odin is needed.
4. This ticket is only for creating/modifying/deleting placeholders. No need to test placeholders resolving in cards for now (separate ticket exists for this use case).
5. Don't use Figma to compare the design, the current version is work in progress and has many things that we decided are not needed and should be removed (including status column). When a placeholder is created it is automatically published making "status" trivial.

**Comment by Axelcureno** [comment]
I have addressed all of your last comments @yesil

**Comment by Roycethan** [comment]
> @Roycethan my notes regarding testing of placeholders.
> 
> 1. I have removed offer selector tool from placeholder creation, that is not in scope for now.
> 2. I was able to modify links on an existing placeholder (added/removed link, changed style) with no issues, please try again and if the issue persists let me know what placeholder you are trying with.
> 3. Please test only in locales that already have placeholders for now. This is happening because not all locales have the folder created in Odin. @npeltier said this will be handled separately in another ticket, a script to create the folders in Odin is needed.
> 4. This ticket is only for creating/modifying/deleting placeholders. No need to test placeholders resolving in cards for now (separate ticket exists for this use case).
> 5. Don't use Figma to compare the design, the current version is work in progress and has many things that we decided are not needed and should be removed (including status column). When a placeholder is created it is automatically published making "status" trivial.

@Axelcureno  For edit link, try removing link from the below placeholder ( by clicking Remove link) and save (at right) and see link still exists:
image

File: studio/src/mas-fragment.js [hunk 39]
```diff
@@ -87,4 +87,4 @@ class MasFragment extends LitElement {
     }
 }
 
-customElements.define('mas-fragment', MasFragment);
+customElements.define('mas-fragment', MasFragment);
```
Comment (yesil) [inline studio/src/mas-fragment.js:90]: EOF

File: studio/src/mas-content.js [hunk 40]
```diff
@@ -27,11 +28,32 @@ class MasContent extends LitElement {
...
+                this.requestUpdate();
+            })
+        );
     }
 
     disconnectedCallback() {
         super.disconnectedCallback();
         Events.fragmentAdded.unsubscribe(this.goToFragment);
+
+        if (this.subscriptions && this.subscriptions.length) {
```
Comment (yesil) [inline studio/src/mas-content.js:49]: please prefer ReactiveController for pure refresh purposes.

File: studio/src/mas-placeholders.js [hunk 41]
```diff
@@ -0,0 +1,2068 @@
...
+        this.folderData = [];
+        this.foldersLoaded = false;
+        this.placeholdersData = [];
+        this.placeholdersLoading = false;
+        this.isBulkDeleteInProgress = false;
+
+        this.reactiveController = new ReactiveController(this, [
+            Store.search,
+            Store.filters,
+            Store.folders?.data,
```
Comment (yesil) [inline studio/src/mas-placeholders.js:84]: folders is always defined. 
`property?` should be used only for props that can be set/unset later.

File: studio/src/mas-placeholders.js [hunk 42]
```diff
@@ -0,0 +1,2068 @@
...
+export function getDictionaryPath(folderPath, locale) {
+    if (!folderPath || !locale) return null;
+    return `${ROOT_PATH}/${folderPath}/${locale}/dictionary`;
+}
+
+class MasPlaceholders extends LitElement {
+    static styles = styles;
+
+    static properties = {
+        searchQuery: { type: String, state: true },
```
Comment (yesil) [inline studio/src/mas-placeholders.js:28]: 👍

File: studio/src/mas-placeholders.js [hunk 43]
```diff
@@ -0,0 +1,2068 @@
...
+                }
+            });
+
+            currentFields.forEach((field) => {
+                if (!fieldsToUpdate.some((f) => f.name === field.name)) {
+                    savePayload.fields.push({ ...field });
+                }
+            });
+
+            const saveResponse = await fetch(fragmentEndpoint, {
```
Comment (yesil) [inline studio/src/mas-placeholders.js:455]: why not use, saveFragment from `mas-repository`

File: studio/src/mas-placeholders.js [hunk 44]
```diff
@@ -0,0 +1,2068 @@
...
+     * @returns {Promise<boolean>} - Success status
+     */
+    async createIndexFragment(parentPath, fragmentPath) {
+        if (!parentPath || !fragmentPath) {
+            return false;
+        }
+
+        try {
+            const repository = this.repository;
+            if (!repository) {
```
Comment (yesil) [inline studio/src/mas-placeholders.js:1767]: I think this should be used only in the render method. 
All the dependencies that are needed, should be pre-conditions to render placeholders page.
Otherwise, this can of checks make the code only more verbose.

```
            if (!repository) {
                return false;
            }
```

File: studio/src/mas-repository.js [hunk 45]
```diff
@@ -90,6 +88,10 @@ export class MasRepository extends LitElement {
         this.style.display = 'none';
     }
 
+    disconnectedCallback() {
```
Comment (yesil) [inline studio/src/mas-repository.js:91]: this is not needed.

File: studio/src/mas-placeholders.js [hunk 46]
```diff
@@ -0,0 +1,2068 @@
...
+                            'Failed to publish index, but update was successful',
+                        );
+                    }
+                    return true;
+                }
+                return false;
+            } catch (error) {
+                if (error.message?.includes('412')) {
+                    await new Promise((resolve) => setTimeout(resolve, 500));
+                    return await this.updateIndexFragment(
```
Comment (yesil) [inline studio/src/mas-placeholders.js:1030]: this can cause a loop forever.

**Review by yesil — COMMENTED** [review]
no blocker comments, but certains things I already commented like using fetch only from aem.js, is still not fully implemented. I'm ok to merge if no regression though.
those can be addressed later.

**Comment by Roycethan** [comment]
> > @Roycethan my notes regarding testing of placeholders.
> > 
> > 1. I have removed offer selector tool from placeholder creation, that is not in scope for now.
> > 2. I was able to modify links on an existing placeholder (added/removed link, changed style) with no issues, please try again and if the issue persists let me know what placeholder you are trying with.
> > 3. Please test only in locales that already have placeholders for now. This is happening because not all locales have the folder created in Odin. @npeltier said this will be handled separately in another ticket, a script to create the folders in Odin is needed.
> > 4. This ticket is only for creating/modifying/deleting placeholders. No need to test placeholders resolving in cards for now (separate ticket exists for this use case).
> > 5. Don't use Figma to compare the design, the current version is work in progress and has many things that we decided are not needed and should be removed (including status column). When a placeholder is created it is automatically published making "status" trivial.
> 
> @Axelcureno For edit link, try removing link from the below placeholder ( by clicking Remove link) and save (at right) and see link still exists: image
> 

Also the text boxes can be made in line in one row 
image
